### PR TITLE
feat(j-s): Prosecutor Asks for Cancellation

### PR DESCRIPTION
**Pull request by gudjong**
# Prosecutor Asks for Cancellation

[Afturköllun ákæru - takki hjá sækjanda](https://app.asana.com/0/1199153462262248/1207421173939600/f)

## What

- Allows prosecutors to ask for the cancellation of indictments after sending them to a district court and before the court has held proceedings for the case.
- This PR only handles the prosecutor ask. Notifications and subsequent court handling remain to be implemented.

## Why

- Part fo indictment flow.

## Screenshots / Gifs

image
image

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
  - Introduced a new case state `WAITING_FOR_CANCELLATION` and transition `ASK_FOR_CANCELLATION`.
  - Added a new modal for requesting indictment cancellation in the prosecutor's overview.
  - New prop `hideActionButton` added to the `FormFooter` component to control action button visibility.

- **Bug Fixes**
  - Improved state handling during case creation and transitions to ensure consistent case state assignments.

- **Improvements**
  - Enhanced user access checks and query filters to include the new `WAITING_FOR_CANCELLATION` state.
  - Updated UI components to reflect new case states and transitions with appropriate messages and styles.

### CODE REVIEW

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Overview/Overview.tsx [hunk 1]
```diff
@@ -239,6 +255,21 @@ const Overview: React.FC<React.PropsWithChildren<unknown>> = () => {
           nextIsDisabled={
             userCanSendCaseToCourt && !indictmentConfirmationDecision
           }
+          hideActionButton={
+            workingCase.state === CaseState.WAITING_FOR_CANCELLATION
+          }
+          actionButtonText={formatMessage(strings.askForCancellationButtonText)}
+          actionButtonColorScheme={'destructive'}
```
Comment (oddsson) [inline apps/judicial-system/web/src/routes/Prosecutor/Indictments/Overview/Overview.tsx:262]:
```suggestion
          actionButtonColorScheme="destructive"
```

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Overview/Overview.tsx [hunk 2]
```diff
@@ -239,6 +255,21 @@ const Overview: React.FC<React.PropsWithChildren<unknown>> = () => {
...
+          actionButtonIsDisabled={
+            !(
```
Comment (oddsson) [inline apps/judicial-system/web/src/routes/Prosecutor/Indictments/Overview/Overview.tsx:264]: This code seems a bit hard to read. Consider extracting this to a function instead, something like

```
const isActionButtonDisabled = () => {
    const isSubmittedOrReceived = [CaseState.SUBMITTED, CaseState.RECEIVED].includes(workingCase.state)
    const hasIndictmentDecision = workingCase.indictmentDecision
    const hasState = workingCase.state
    
    return isSubmittedOrReceived && hasIndictmentDecision && !hasState 
}
```

Check the actual implementation though 😅

**Review by oddsson — COMMENTED** [review]
A couple of small consideration tasks

File: apps/judicial-system/backend/migrations/20240531113155-update-case.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+          { transaction },
+        )
+        .then(() =>
+          queryInterface.sequelize.query(
+            `ALTER TABLE "case" ALTER COLUMN state SET DEFAULT 'NEW'`,
+            { transaction },
+          ),
+        ),
+    )
+  },
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/migrations/20240531113155-update-case.js:52]: Keep it implicit.

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Overview/Overview.tsx (hunk 2 above)
Comment (gudjong) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Overview/Overview.tsx:264]: I used a different approach and introduced a variable with a descriptive name.

**Review by oddsson — APPROVED** [review]

**Pull request merged**